File: HeadFirstDesignPatternsOC/StrategyPattern/Duck/Duck.m
//
//  Duck.m
//  StrategyPattern
//
//  Created by 陳峻琦 on 25/5/2022.
//

#import "Duck.h"

@implementation Duck

- (void)swim {
    NSLog(@"All ducks float, even decoys!");
}

- (void)display {
    [NSException raise:NSInternalInconsistencyException format:@"You must override %@ in a subclass", NSStringFromSelector(_cmd)];
}

- (void)performFly {
    [self.flyBehavior fly];
}

- (void)performQuack {
    [self.quackBehavior quack];
}

@end

File: HeadFirstDesignPatternsJava/HeadFirst/src/StrategyPattern/Duck/Duck.java
package StrategyPattern.Duck;

import StrategyPattern.Fly.FlyBehavior;
import StrategyPattern.Quack.QuackBehavior;

public abstract class Duck {
    protected FlyBehavior flyBehavior;
    protected QuackBehavior quackBehavior;

    public Duck() {

    }

    public abstract void display();

    public void swim() {
        System.out.println("All ducks float, even decoys!");
    }

    //委托给行为类
    public void performFly() {
        flyBehavior.fly();
    }

    //委托给行为类
    public void performQuack() {
        quackBehavior.quack();
    }

    public void setFlyBehavior(FlyBehavior fb) {
        flyBehavior = fb;
    }

    public void setQuackBehavior(QuackBehavior qb) {
        quackBehavior = qb;
    }
}
